skiko/src/jvmMain/kotlin/org/jetbrains/skiko/Library.kt

```diff
@@ -42,27 +47,37 @@ object Library {
         return file
     }
 
-    private var loaded = AtomicBoolean(false)
+    /**
+     * Holds a reference to the lock which has to be acquired when loading the native library,
+     * or to wait for the loading to finish.
+     * The reference will resolve to `null` if loading is done and callers do not need to wait anymore.
+     */
+    private val loadingLock = AtomicReference(ReentrantLock())
 
     // This function does the following: on request to load given resource,
     // it checks if resource with given name is found in content-derived directory
     // in Skiko's home, and if not - unpacks it. It could also load additional
     // localization resources, on platforms where it is needed.
-    @Synchronized
     fun load() {
-        if (!loaded.compareAndSet(false, true)) return
-
-        // Find/unpack a usable copy of the native library.
-        findAndLoad()
-
-        // TODO move properties to SkikoProperties
-        Setup.init()
-
-        try {
-            // Init code executed after library was loaded.
-            org.jetbrains.skia.impl.Library._nAfterLoad()
-        } catch (t: Throwable) {
-            t.printStackTrace()
+        val lock = loadingLock.get() ?: return
+        lock.withLock {
+            // We entered the critical section, but another thread might have already entered and finished
+            if (loadingLock.get() !== lock) return
```

Comment:
I debugged this code, and this lock doesn't work - `load()` is triggered recursively from `org.jetbrains.skia.impl.Library._nAfterLoad()` in the same thread.

The previous `Synchronized` + `!loaded.compareAndSet(false, true)` seems worked as expected?

Reply:
> No, as mentioned in the commit message, the Synchronized + Atomic Boolean guard is suspected of returning too early

It looked to me that only `staticLoad` had the issue, the `load` itself was protected by `Synchronized` which shouldn't allow returning too early, it should block if another thread enters it.

The new code looks like it has the same logic to me:
- instead of `Synchronized` we use `ReentrantLock`
- instead of setting a boolean to false in the beginning we set the lock to null in the end, and add the `isHeldByCurrentThread` check

It is okay to keep it, but the initial logic looked slightly simpler to me.